toolchain: Add GCC 9.1.0 release

jbenden opened the pull request:
Add GCC 9.1.0 (just released last Friday) to OpenWrt toolchain.

Tested against: master branch, with mips_24kc_musl

jbenden committed:
toolchain: Add GCC 9.1.0 release

Signed-off-by: Joseph Benden <joe@benden.us>

diizzyy commented:
Nice, patch 950 seems to be already fixed as far as I can tell?

https://gcc.gnu.org/bugzilla/show_bug.cgi?id=47047 --> https://gcc.gnu.org/bugzilla/show_bug.cgi?id=70268

wongsyrone commented:
950 is replaced by another option

https://github.com/openwrt/openwrt/commit/139f99c05880508b19308eed366d8fbf5804fbf9#diff-ea5d5eec1b3a6d00baf5414fc2aadfe3

BTW, here is mine https://github.com/wongsyrone/lede-1/commit/288e853241811a00241e94dcbce4cc122738568e

jbenden commented:
Thanks for the feedback!

wongsyrone commented:
There are two commits in the upstream gcc-9-branch worth to be backported, any idea?

https://github.com/gcc-mirror/gcc/commit/f36b864c7d84b36883c6190c83b31c0a8c15172b

https://github.com/gcc-mirror/gcc/commit/7a1606168f60622f73a7dd90778e2a148a2c520c

wongsyrone commented:
BTW, please refresh patches

wongsyrone commented:
Can you test gcc 9 + binutils 2.32 for x86_64? I cannot build e2fsprogs, it complains libuuid.so.1 not found and undefined references.

jbenden commented:
I have refreshed the patches, as well as added the two additional patches.

I will research the `x86_64` problem later today, hopefully.

wongsyrone commented on toolchain/gcc/Config.in:
```diff
@@ -17,6 +17,10 @@ choice
+
+	config GCC_USE_VERSION_9
+		bool "gcc 9.x"
+		depends on !arc
```
It seems arc is available since gcc 8.3 thus no need to exclude gcc 9

https://github.com/openwrt/openwrt/pull/1803#issuecomment-462334890
https://github.com/openwrt/openwrt/commit/343b36ffc79f9506586b5263b5d840dce0d801e0

jbenden replied:
I pushed the change up. Thanks!!

jbenden commented:
During research in to the problem report of `e2fsprogs`, I fixed a few issues related to compiler warnings and/or errors.

Should I include these fixes in this PR? Or should another PR be opened for each reported issue?

Namely, the patches relate to the following:

1. ubox `kmodloader.c`: the buffer must be larger than the possible copied string; the fix is to increase line 33's `path` to 1024.
2. `util-linux` must have the pkg-config file for `libblkid.so` adjusted to always include `libuuid.so`.

@wongsyrone The problem you experience is resolved by issue 2, above.

wongsyrone commented:
Same PR but in different commits. I'm still confused, why all things work well when using gcc 8.3 + binutils 2.32.

jbenden commented:
The actual problem with `e2fstools` is that during compilation of the binaries; it references `libblkid.so` but not `libuuid.so`. On the otherhand, it SEEMS that `libblkid.so` references `libuuid.so`. But I cannot be sure as my `ldd` won't display the shared libraries loaded...

jbenden commented:
Is `util-linux` where your version of `libblkid.so` and `libuuid.so` originate from?

wongsyrone commented:
I tried both util-linux 2.33 and 2.33.2 with no luck.

BTW, I have linking issues for iproute2 as well, it cannot find zlib when using tc/ip-full variant, libelf linked to libz, looking the same scenario as e2fsprogs, several commits in my tree to fix gcc9 build as below

https://github.com/wongsyrone/lede-1/commit/4ef019a56bbc1939285159d223ff43c1500b0158
https://github.com/wongsyrone/lede-1/commit/7c4a16bbc52ec4a762ca5eb32f37be5b5514794c
https://github.com/wongsyrone/lede-1/commit/1e679625528795b9251e19adca78200d16fae350
https://github.com/wongsyrone/lede-1/commit/4adf07c2f6393e38b86a153e4fb1c6838f27bbf9

As stated before, all issues disappear when using gcc 8.3 + binutils 2.32 even for the original version in openwrt tree

jbenden commented:
It may be LTO, based on initial research.

jbenden commented:
I tried using the previous version of binutils; it is also broken.

I explicitly disabled GCC final's `--enable-lto`; it is still broken.

So the problem is somehow related to just GCC 9.1, and does not seem to be related to LTO.

wongsyrone commented:
@nbd168 Can you help to debug GCC 9.1 issues?

jbenden commented:
I believe I've fixed the linking problem. @wongsyrone Could you please test?

I still have at least one additional commit to add, once I know this part is proper... Thanks!

hauke commented on toolchain/gcc/Config.version:
```diff
@@ -7,13 +7,20 @@ config GCC_VERSION_8
 
+config GCC_VERSION_9
+	default y if GCC_USE_VERSION_9
+	default y if arc
```
Please do not make this default on arc

jbenden replied:
I was asked to add this. So should it be removed?

jbenden replied:
I've remove it.

hauke commented on scripts/patch-specs.sh:
```diff
@@ -37,6 +37,9 @@ patch_specs() {
 						sub("%{L.}", "%{L*} -L %:getenv(STAGING_DIR /usr/lib) -rpath-link %:getenv(STAGING_DIR /usr/lib)")
 					}
+					mode ~ "link" {
+						sub("%@{L.}", "%@{L*} -L %:getenv(STAGING_DIR /usr/lib) -rpath-link %:getenv(STAGING_DIR /usr/lib)")
```
https://patchwork.ozlabs.org/patch/1098467/

jbenden replied:
huh?

jbenden replied:
Error?

hauke replied:
I wanted to ask why is this needed ;-)

wongsyrone replied:
@hauke 

You can try without the patch-specs.sh change and see what happens when compiling packages especially e2fsprogs and iproute2 on x86_64 and may fix in a reasonable way. BTW, you commented on the outdated change.

jbenden committed:
package/system/ubox: Increase variable path's buffer size

This corrects for an error during compilation, whereby the stack
variable `path` is not long enough for the snprintf on line 95.
GCC reports the snprintf output is between 11 and 330 bytes.

Signed-off-by: Joseph Benden <joe@benden.us>

jbenden commented:
With these patches, both `x86-64_gcc-9.1.0_musl` and `mips_24kc_musl` test OK.

wongsyrone commented:
The only difference is the change on scripts/patch-specs.sh? Can you explain it?

jbenden commented:
It modifies GCC's spec file; by adding multiple `getenv` calls to acquire the `STAGING_DIR` variable, during compilation. This changes where the linker and compiler look for additional includes and libs.

jbenden commented:
I've simplified the AWK expression; hoping to have combined both flavors correctly.

wongsyrone commented on toolchain/gcc/Config.version:
```diff
@@ -7,13 +7,19 @@ config GCC_VERSION_8
+
 config GCC_USE_EMBEDDED_PATH_REMAP
 	default y if GCC_VERSION_8
+	default y if GCC_VERSION_9
```
I suggest using `default y if ( GCC_VERSION_8 || GCC_VERSION_9 )` to combine these two lines

jbenden replied:
Done!

wongsyrone commented:
Compiled tested on x86_64 using the previous AWK expression.

wongsyrone commented:
I think it's ready for @hauke  to review as you have push access.

neheb commented:
I tested this. Compiles fine. Was expecting build failures.

dengqf6 commented:
Works fine on mvebu with binutils 2.32

InBetweenNames commented:
Tested, working great on mvebu, WRT3200ACM, musl libc.

hingbong commented:
working well on r7800 linux 4.19 with binutils 2.32

EXSERENS commented:
Test passed on WRT3200ACM with kernel 4.19 and bintuil 2.32.

Ansuel commented:
Working well on wrt3200acm and r7800

hauke commented:
Thank you for the patch, it was merged into master.

borkra commented:
Please fix this uhttpd with gcc 9.1

The problem is these lines in utils.c:

ustream_printf(cl->us, "\r\n", len);

neheb commented:
What's wrong with it?

borkra commented:
No format specifier for len ...

Here is a patch I had to create to get all problems of uhttpd with gcc 9.1 fixed:

-- a/utils.c	2019-06-16 12:22:14.477449301 -0400
+++ b/utils.c	2018-11-28 06:36:35.000000000 -0500
@@ -47,7 +47,7 @@
 		ustream_printf(cl->us, "%X\r\n", len);
 	ustream_write(cl->us, data, len, true);
 	if (chunked)
-		ustream_printf(cl->us, "\r\n", len);
+		ustream_printf(cl->us, "\r\n");
 }
 
 void uh_chunk_vprintf(struct client *cl, const char *format, va_list arg)
@@ -74,7 +74,7 @@
 		ustream_write(cl->us, buf, len, true);
 	else
 		ustream_vprintf(cl->us, format, arg);
-	ustream_printf(cl->us, "\r\n", len);
+	ustream_printf(cl->us, "\r\n");
 }
 
 void uh_chunk_printf(struct client *cl, const char *format, ...)
--- a/proc.c	2018-11-28 06:36:35.000000000 -0500
+++ b/proc.c	2019-06-16 12:30:56.077271848 -0400
@@ -232,7 +232,7 @@
 	uloop_timeout_cancel(&p->timeout);
 	uh_http_header(cl, cl->dispatch.proc.status_code, cl->dispatch.proc.status_msg);
 	blob_for_each_attr(cur, cl->dispatch.proc.hdr.head, rem)
-		ustream_printf(cl->us, "%s: %s\r\n", blobmsg_name(cur), blobmsg_data(cur));
+		ustream_printf(cl->us, "%s: %s\r\n", blobmsg_name(cur), (char*)blobmsg_data(cur));
 
 	ustream_printf(cl->us, "\r\n");
 
--- a/ubus.c	2018-11-28 06:36:35.000000000 -0500
+++ b/ubus.c	2019-06-16 12:38:42.515168641 -0400
@@ -144,11 +144,11 @@
 	}
 
 	ustream_printf(cl->us, "Access-Control-Allow-Origin: %s\r\n",
-	               blobmsg_data(tb[HDR_ORIGIN]));
+	               (char*)blobmsg_data(tb[HDR_ORIGIN]));
 
 	if (tb[HDR_ACCESS_CONTROL_REQUEST_HEADERS])
 		ustream_printf(cl->us, "Access-Control-Allow-Headers: %s\r\n",
-		               blobmsg_data(tb[HDR_ACCESS_CONTROL_REQUEST_HEADERS]));
+		               (char*)blobmsg_data(tb[HDR_ACCESS_CONTROL_REQUEST_HEADERS]));
 
 	ustream_printf(cl->us, "Access-Control-Allow-Methods: POST, OPTIONS\r\n");
 	ustream_printf(cl->us, "Access-Control-Allow-Credentials: true\r\n");

dengqf6 commented:
When testing on ramips, I found luci no longer works because of liblua failure:
```
[  186.660860] do_page_fault(): sending SIGSEGV to luci for invalid read access from 00000008
[  186.669368] epc = 77e83ee3 in liblua.so.5.1.5[77e76000+2d000]
[  186.675252] ra  = 77e83f7f in liblua.so.5.1.5[77e76000+2d000]
```
Compiler flags:
`-Os  -pipe -mno-branch-likely -march=24kec`
Replace `Os` with `O0` or `O2` will make it work again.

Would you look into this?

neheb commented:
Now I know that it's GCC9 that I can't access LuCI :)

wongsyrone commented:
What? It works well on my x86_64 box.

neheb commented:
He's talking about ramips.

dengqf6 commented:
@neheb And then I found zerotier always fails, regardless of optimization level.
```
[  375.218646] do_page_fault(): sending SIGSEGV to zerotier-one for invalid read access from 3e55328c
[  375.227865] epc = 00408de9 in zerotier-one[400000+db000]
[  375.233303] ra  = 004144ab in zerotier-one[400000+db000]
```

neheb commented:
@dengqf6 is it only lua? I'm getting that error even when filtering Os flags.

hauke commented:
Do you still have these problem with gcc 9.2?

neheb commented:
Actually...I don't think my firmware is built using 9.2. I will retest.

neheb commented:
@dengqf6 do you have this issue? It seems to have resurfaced for me: https://gist.github.com/neheb/903d07b5f7b7e9796e026bafe819d68c
